app_streamlit.py

```diff
@@ -1,78 +1,71 @@
+    ax.set_title("Contagem de Nucleotídeos")
+    ax.set_xlabel("Nucleotídeos")
+    ax.set_ylabel("Quantidade")
+    ax.grid(True)
+
+    st.pyplot(fig)
+
 def main():
     st.title("Contador de Nucleotídeos de DNA")
-
-    #upload arquivo
     uploaded_file = st.file_uploader("Faça o upload de um arquivo de texto com a sequência de DNA", type=["txt"])
-
-    # armazenena sequência de DNA
     dna_sequence = ""
 
-    #se der upload, processa
     if uploaded_file is not None:
-        dna_sequence = processar_arquivo(uploaded_file)
+        dna_sequence = process_uploaded_file(uploaded_file)
         st.write("Sequência de DNA do arquivo:")
-        st.text(dna_sequence)  # Exibe a sequência do arquivo
-
-    #sem upload
+        st.text(dna_sequence)
     else:
         dna_sequence = st.text_input("Ou digite uma sequência de DNA:").upper()
 
-    #valida
-    if dna_sequence:
-        if validar_dna(dna_sequence):
-            nucleotides_count = {
-                "A": dna_sequence.count("A"),
-                "T": dna_sequence.count("T"),
-                "C": dna_sequence.count("C"),
-                "G": dna_sequence.count("G")
-            }
-
-            #porcentagem
-            st.write("### Porcentagem de cada nucleotídeo:")
-            for nucleotide, count in nucleotides_count.items():
-                percentage = (count / len(dna_sequence)) * 100
-                st.write(f"{nucleotide}: {percentage:.2f}%")
-
-            # GC content
-            gc_content = (nucleotides_count["G"] + nucleotides_count["C"]) / len(dna_sequence) * 100
-            st.write(f"### GC Content: {gc_content:.2f}%")
-
-            # AT content
-            at_content = (nucleotides_count["A"] + nucleotides_count["T"]) / len(dna_sequence) * 100
-            st.write(f"### AT Content: {at_content:.2f}%")
-
-            # grafico
-            nucleotides = list(nucleotides_count.keys())
-            counts = list(nucleotides_count.values())
-
-            fig, ax = plt.subplots()
-            ax.bar(nucleotides, counts, color=['blue', 'orange', 'green', 'red'])
-            ax.set_title("Contagem de Nucleotídeos")
-            ax.set_xlabel("Nucleotídeos")
-            ax.set_ylabel("Quantidade")
-            ax.grid(True)
-
-            st.pyplot(fig)  
-
-        else:
-            st.error("Sua sequência contém caracteres inválidos! Apenas A, T, C e G são permitidos.")
+
+    if not is_valid(dna_sequence):
+        st.error("Sua sequência contém caracteres inválidos! Apenas A, T, C e G são permitidos.")
+        return
+
+    if not dna_sequence:
+        return
+
+    nucleotides_count = Counter(dna_sequence)
+    dna_sequence_size = len(dna_sequence)
+
+    write_nucleotides_percentage(nucleotides_count, dna_sequence_size)
+    write_gc_content(nucleotides_count, dna_sequence_size)
+    write_at_content(nucleotides_count, dna_sequence_size)
+    plot_nucleotides_count_graph(nucleotides_count)
```

Comment:
Separated into different reusable functions with a single responsibility. No further comment is needed as the fn name tells the whole story.